docs/TrafficSplitting.md

```diff
@@ -0,0 +1,120 @@
+# Traffic Splitting
+
+The purpose of this section of the kn documentation is to describe the command verbs for traffic splitting operations and present some examples.
+
+`kn` client can work with following properties of a traffic target defined in traffic block of a service spec.
+ - **RevisionName**: RevisionName of a specific revision to which to send a portion of traffic.
+ - **Tag**: Tag is optionally used to expose a dedicated url for referencing this target exclusively.
+ - **Percent**: Percent specifies percent of the traffic to this Revision.
+
+## Operations
+
+Operations required to perform traffic splitting use cases, for example: blue/green deployment, etc can be summarized as below:
+ - Tag targets
+ - Split traffic
+
+### Tag Targets
+
+User can tag traffic targets (revisions) to
+ - Create custom URLs
+ - Create a human-friendly reference to refer to a traffic target
+ - Create static routes and change the underlying revisions as needed
+ - Remove the custom URLs OR remove the tags of targets
+ - Remove a target from traffic block altogether
+
+
+#### Tagging commands
+
+#### 1. `kn service tag-target [RevisionName1:Tag1] [RevisionName2:Tag2]`
```

Comment:
It's not clear to me how this works when there could be more than one traffic block with the specified revisionName. How does this work when there is no revisionName on the block - just `latestRevision:true`? Asking the user to look at the status section to find the current revisionName in that block will be cumbersome.

Reply:
And checking the status section is racey

Reply:
> It's not clear to me how this works when there could be more than one traffic block with the specified revisionName.

You can repeat the same revision name multiple times

> How does this work when there is no revisionName on the block - just `latestRevision:true`

True, that might be an issue. So I would recommend a special revision name for specifying the target referenced with latestResvision. Something which is otherwise an illegal revision name, like "$" or "+".

BTW, a user can always find out all revision names attached to a service via `kn service describe`

Reply:
> >    It's not clear to me how this works when there could be more than one traffic block with the specified revisionName.

> You can repeat the same revision name multiple times

I wasn't clear... the same revisionName might appear on 5 traffic blocks but I want to set a tag on just one of them....how do I specify that one of the 5?

Reply:
> It's not clear to me how this works when there could be more than one traffic block with the specified revisionName.

As Roland mentioned, just repeat the same revision multiple times.

> How does this work when there is no revisionName on the block - just latestRevision:true

I think referencing the entry in traffic block having `latestRevsion:true` is technically confusing, as with every update to service (with `traffic-tag` or `set-traffic` commands) there will be a new revision generated, so which revision should be referenced as `latestRevision` ? The one which was there before issuing the command OR the one that will be created after issuing the command?
I think here, we want the traffic block entry referencing the most latest revision for service (after issuing the current command), and for that we need to reference that via a word or character, which should translate to `latestRevision:true` field in traffic block, but not use the `latestRevision` as is in the CLI (because that could be confusing).

Reply:
> I wasn't clear... the same revisionName might appear on 5 traffic blocks but I want to set a tag on just one of them....how do I specify that one of the 5?

@duglin 
I think you are saying we need a service reference in that tag command ?

Reply:
hmmm I wasn't but I think you might be right about that. Should the service name appear in the cmd ? Or are you assuming that the prefix of the revision name can be used to find the service?  Not sure it can if there are dashes in the service name or revision postfix string. Even if it could I just wonder if that's a confusing UX since most people might expect the cmd to be: `kn service NAME tag-target ...`. This might be where we could use just the unique part of the revision name instead of the full revision name, like we talked about in another thread.

e.g. `kn service echo tag-target v1:sometag` instead of
`kn tag-target echo-v1:sometag` or
`kn service echo tag-target echo-v1:sometag`

Reply:
> are you assuming that the prefix of the revision name can be used to find the service? 

No. We can find the service name of the revision by lookup in revision's labels as
https://github.com/knative/client/blob/9513dedfbae45d38df80ff903da67f569b926c8d/pkg/kn/commands/revision/human_readable_flags.go#L58

this is how we list the service for revision in `kn revision list` command and thats why `kn service tag-target` command was proposed with requiring to refer a service name as we can always figure it out.

Reply:
ah I see. I do wonder though if from a user's perspective that's confusing because the main resource we're trying to get them to work on is a "service" and yet we don't ask them to provide that. Yes we may not need it, but will a user need it in their mental model they're dealing with. I don't ever expect to expose my users to "revisions".  Just something to think about.

Reply:
Sure, if we see the `kn service set-traffic` command, it requires a service name. We could just bring the same experience for `kn service tag-target` command, i.e. requiring the service name as first positional argument (this also saves the lookup for service name and we can then verify that requested revisions to be tagged belongs to the service requested to operate on). 

@rhuss : WDYT about requiring a service name for `kn service tag-target` command as first positional argument ?

Reply:
According to our CLI conventions spec we should *always* use the name of the _noun_ entity as the first argument, so +1 for using the service name here.

I'm wondering whether we could shorten `tag-target` to `tag` as I think `target` does not add much info in addition (and `tag` then just means tag within the route)

Reply:
> although there is `AGE` column but we can ease the lookup by printing the numbers.

:+1: I will update the PR for requiring service name for `tag-target`.

> I'm wondering whether we could shorten tag-target to tag as I think target does not add much info in addition (and tag then just means tag within the route)

IMO, `tag-target` is more explicit specifying its the tag for the traffic target and not the tag for service itself.  Also `tag` word is used to label the resources at other places for example ` knctl`, where `knctl revision tag ` will **label** the revision object as specified. We also have an open issue https://github.com/knative/client/issues/232 on the same line.

Reply:
I have updated the PR to require `serviceName` as first argument for tag and untag commands.